platform/migration/platform-migratable-api/src/main/java/org/codice/ddf/migration/MigrationReport.java

```diff
@@ -0,0 +1,219 @@
+     * @param args   the arguments to the format message
+     * @return this for chaining
+     * @throws IllegalArgumentException if <code>format</code> is <code>null</code>
+     */
+    public default MigrationReport record(String format, @Nullable Object... args) {
+        return record(new MigrationInformation(format, args));
+    }
+
+    /**
+     * Records a message that occurred during the migration report.
+     * <p>
+     * Recorded errors (i.e. MigrationException} will be thrown back when {@link #verifyCompletion()}
+     * is called at the end of the operation.
+     *
+     * @param msg the message to record
+     * @return this for chaining
+     * @throws IllegalArgumentException if <code>msg</code> is <code>null</code>
+     */
+    public MigrationReport record(MigrationMessage msg);
+
+    /**
+     * Registers code to be invoked at the completion of the migration operation.
+     *
+     * @param code the code to execute which will receive this report in parameter where
+     *             additional errors and/or warnings can be registered
+     * @return this for chaining
+     * @throws IllegalArgumentException if <code>code</code> is <code>null</code>
+     */
+    public MigrationReport doAfterCompletion(Consumer<MigrationReport> code);
+
+    /**
+     * Retrieves all messages recorded by the operation that generated this migration report.
+     *
+     * @return a stream of all recorded messages (may be empty)
+     */
+    public Stream<MigrationMessage> messages();
+
+    /**
+     * Retrieves all errors recorded by the operation that generated this migration report.
+     *
+     * @return a stream of all recorded errors (may be empty)
+     */
+    public default Stream<MigrationException> errors() {
+        return messages().filter(MigrationException.class::isInstance)
+                .map(MigrationException.class::cast);
+    }
+
+    /**
+     * Retrieves all warnings recorded by the operation that generated this migration report.
+     *
+     * @return a stream of all recorded warnings (may be empty)
+     */
+    public default Stream<MigrationWarning> warnings() {
+        return messages().filter(MigrationWarning.class::isInstance)
+                .map(MigrationWarning.class::cast);
+    }
+
+    /**
+     * Retrieves all informational messages recorded by the operation that generated this migration
+     * report.
+     *
+     * @return a stream of all recorded info messages (may be empty)
+     */
+    public default Stream<MigrationInformation> infos() {
+        return messages().filter(MigrationInformation.class::isInstance)
+                .map(MigrationInformation.class::cast);
+    }
+
+    /**
+     * Checks if the operation that generated this migration report was successful or not.
+     * <p>
+     * <i>Note:</i> A successful operation might still report warnings.
+     * <p>
+     * Invoking this method will call all registered code via {@link #doAfterCompletion(Consumer)} first.
```

Comment:
❓ It isn't clear to me how the `doAfterCompletion` method is used and works. Since many methods mention that the registered code will be called in their comments, doesn't that imply that the code registered using `doAfterCompletion` could be called multiple times? For instance, if a client calls `wasSuccessful()` and `hasInfo()`, wouldn't the registered code be called twice?

Related to @Lambeaux's comment above, couldn't we add a `doAfterCompletion()` method on `Migratable` that takes a `MigrationReport` and doesn't do anything by default and let `Migratable` implementors overwrite the method if they want to do something when everthing is done? That should allow the framework to guarantee that the method is called only once.

Thoughts?

Reply:
Let me answer each comments ...
1) I will update the documentation to indicate the fact that a registered code will only be executed once before a successful result is returned. I will remove the notes in all methods that indicate the registered codes will be first executed.
2) As for moving it to the Migratable, I believe this would complicate the life of the Migratable implementor. If one needs to register code based on various paths during the store(), they would need to keep a list of all code to invoke themselves when the framework finally calls them. Keeping this way, simplifies the migrtable's logic.